File: src/main/java/com/kasisoft/libs/common/functional/KBiFunction.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KBiFunction<T, U, R> {

    R apply(T input1, U input2) throws Exception;

    @NotNull
    default <V> KBiFunction<T, U, V> andThen(@NotNull KFunction<? super R, ? extends V> after) {
        return ($t, $u) -> after.apply(apply($t, $u));
    }

    @NotNull
    default BiFunction<T, U, R> protect() {
        return ($t, $u) -> {
            try {
                return apply($t, $u);
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KBiFunction.kt
package com.kasisoft.libs.common.functional

import com.kasisoft.libs.common.*
import java.util.function.Function
import java.util.function.*

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KBiFunction<T: Any?, U: Any?, R: Any?> {

    @Throws(Exception::class)
    fun apply(t: T, u: U): R

    fun protect(): BiFunction<T, U, R> =
        BiFunction<T, U, R> { t: T, u: U ->
            try {
                apply(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
